android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java

```diff
@@ -512,16 +356,150 @@ public void updateLayout(
                 + parentTag
                 + " as a parent, but its Manager doesn't implement IViewManagerWithChildren");
       }
-      if (parentViewManagerWithChildren != null
-          && !parentViewManagerWithChildren.needsCustomLayoutForChildren()) {
+      if (!parentViewManagerWithChildren.needsCustomLayoutForChildren()) {
         viewToUpdate.layout(x, y, x + width, y + height);
       }
     } else {
       viewToUpdate.layout(x, y, x + width, y + height);
     }
   }
 
+  public boolean hasAnimationForTag(int tag, String type) {
+    return mNativeMethodsHolder.hasAnimationForTag(tag, type);
+  }
+
   public boolean isLayoutAnimationEnabled() {
     return mNativeMethodsHolder != null && mNativeMethodsHolder.isLayoutAnimationEnabled();
   }
+
+  private boolean removeOrAnimateExitRecursive(View view, ViewGroup parent, boolean shouldRemove) {
+    int tag = view.getId();
+    boolean hasExitAnimation = hasAnimationForTag(tag, "exiting");
+    boolean wantAnimateExit = hasExitAnimation;
+
+    // we might want to keep this view around
+    // because one of the (children's) children
+    // has an exiting animation
+    if (view instanceof ViewGroup) {
+      ViewGroup viewGroup = (ViewGroup) view;
+      ArrayList<View> children = new ArrayList<>();
+      for (int i = 0; i < viewGroup.getChildCount(); i++) {
+        View child = viewGroup.getChildAt(i);
+        children.add(child);
+      }
+      for (View child : children) {
+        if (removeOrAnimateExitRecursive(child, viewGroup, shouldRemove && !hasExitAnimation)) {
+          wantAnimateExit = true;
+        }
+      }
+    }
+
+    if (hasExitAnimation) {
+      Snapshot before = new Snapshot(view, mReanimatedNativeHierarchyManager);
+      HashMap<String, Object> currentValues = before.toCurrentMap();
+      HashMap<String, Float> preparedValues = prepareDataForAnimationWorklet(currentValues, false);
+      if (!mExitingViews.containsKey(tag)) {
+        mExitingViews.put(tag, view);
+        registerExitingAncestors(view);
+        mNativeMethodsHolder.startAnimationForTag(tag, "exiting", preparedValues);
+      }
+      return true;
+    }
+
+    if (wantAnimateExit) {
+      mAncestorsToRemove.add(tag);
+      return true;
+    }
+
+    if (shouldRemove) {
+      removeView(view, parent);
+    } else {
+      view.setClickable(false);
+    }
+    return false;
+  }
+
+  private void registerExitingAncestors(View view) {
+    View parent = (View) view.getParent();
+    while (parent != null && !(parent instanceof RootView)) {
+      int tag = parent.getId();
+      Integer previousValue = mExitingSubviewCountMap.get(tag);
```

Comment:
maybe use AtomicInteger such that we can just mutate the value instead of replacing objects in a hasmap (not something v important tho)

e.g.:

```
mExitingSubviewCountMap.putIfAbsent(tag, new AtomicInteger(0)).incrementAndGet()
```

Reply:
`putIfAbsent` complains about Android version support, so I'm not sure if we can use it here.
As for storing a mutable in the map, I don't know. I'd say it's fine as is.